2016/proposals/pravendra-singh.md

```diff
@@ -0,0 +1,220 @@
+
+The following proposal talks down an application which can "*Represent the survey responses in a comprehensible manner, assisting the workshop organizers (hosts) to plan the event according to the feedback from attendees*".
+
+The application will be a relieve from the current cumbersome method we use for data aggregation.
+
+It has an additional benefit here, as we (Software Carpentry) believe in openness, the resultant application can be open sourced for the great good of others who want to utilize the SurveyMonkey survey responses in an interactive fashion.
+
+## Technical Details
+
+The application will be powered by Django, as we also want it to work with AMY. It will use Plotly.js for visualizing different response types, which is a open source [project](https://plot.ly/javascript/open-source-announcement/).
+
+Going by the "*Object-oriented analysis and design*" approach for system design, we can categorize the application in the following sections.
+
+###### 1. Visual representation of the survey questions
+
+Both the type of surveys have multiple questions of varying classes. Ideally, a good visualizer should respect the inherent attributes of the dataset. That's why the application will employ suitable visualization for different types of questions.
+
+According to this [explanation](https://github.com/numfocus/gsoc/issues/77#issuecomment-192500549), here is a list of suitable charts.
+
+| Question Type  | Suitable Chart Type | Sample Question |
+| ------------- | ------------- | ------------- |
+| Multiple choice - Single answer  | Bar and Pie chart  | When are you taking this survey? |
+| Multiple choice - Multiple answer  | Chord diagram  |  What is your domain of research/study? |
+| Yes-No-Maybe type  | Horizontal Stacked Bar chart  |  How would you describe your ability to do the following tasks? |
+| Matrix type  | Heatmap Matrix chart  | How would you rate your motivation to learn about these topics? |
+| Text as an input  | Text-Cloud OR Table OR Semantic Analysis | In a few words, what is your most important reason for attending this workshop? |
+
+###### 2. Response-Cache service
+
+As we are already aware of the order of total responses we get from our attendees, having a "*Data Cache Layer*" is an intelligent solution. Rate Limitations exposed by SurveyMonkey API is another reason behind implementing a cache service.
+
+The *Response-Cache Service* will be responsible for serving cached responses, and hence plots. It will leverage the *offline* mode of Plotly so that the caching service is only dependent on the API responses and idempotent in terms of generated plots.
+
+The service will use *Heartbeat Mechanism* to invalidate cache content with respect to the API response periodically. Workshop identifiers (ID) will be used to collect and invalidate responses for a particular event.
```

Comment:
Hi @pbanaszkiewicz,

[Heartbeat](https://en.wikipedia.org/wiki/Heartbeat_(computing)) is a terminology mostly used in Distributed Systems, to monitor the availability of a resource. Clients ask the central node about a particular resource, and then the main system returns a response after some computation, stating whether it is available or not.

It's called "Heartbeat" because clients generate it *periodically*. Simply putting, we can refer it to the [Long Polling](https://en.wikipedia.org/wiki/Push_technology#Long_polling) technique.

In the proposal, I meant to say that the "Data Aggregation Service" (the one that will collect survey responses) will ask the SurveyMonkey API for responses, *periodically*.

I am updating the details about "Data Aggregation Service" soon.